static/js/util.ts

```diff
@@ -14,11 +16,13 @@ export function random_int(min, max) {
 //
 // Usage: lower_bound(array, value, [less])
 //        lower_bound(array, first, last, value, [less])
-export function lower_bound(array, arg1, arg2, arg3, arg4) {
-    let first;
-    let last;
-    let value;
-    let less;
+type LessFunc<T> = (a: T | number, b: number, c: number) => boolean;
+export function lower_bound<T>(array: T[], arg1: number, arg2?: number | LessFunc<T>,
```

Comment:
I am not 100% happy with the `T | number`  here as the type of the first argument here but not having it here is causing ts error. Even though, this isn't wrong either. For context, the less function here is called with an item in the array passed in which could be a number or an object hence the generic `T` but it also could be undefined where we supply the default less function which assumes that the array is made of numbers.

I tried doing a type narrow by doing this:
```typescript
function is_num_array(arr: unknown[]): arr is number[] {
    if (arr.length === 0) {
          return true; // the function is not needed for empty array
     } 
     return (typeof a[0] === 'number');
}

// in the function under the less is an undefined condition
if (is_num_array(array)) {
     less = (a, b) => a > b;
}
```

But that does not work! Anything I am doing wrong here @andersk?

Reply:
I fixed this in 1df9d28898dc04b1a5adf9766606c3c934608679. The solution here is to add number check in the default `less` function itself. The above type narrowing make these type of the array: `T & number`.